File: apps/desktop/src/lib/curve.ts
// 曲線の折り目(CPE-011)の計算。円弧・3次ベジェを、指定した誤差以内の折れ線にする。
//
// 展開図のデータ構造は直線の辺だけなので、曲線は「十分細かい折れ線」として入れる。
// 同じ計算はRust側(crates/ori3-cp/src/curve.rs)にもある。Tauriコマンドは
// 9個のまま増やさない約束なので、描いている最中の形はここで作り、確定したときに
// 既存の edit_apply(AddSegment)を折れ線の本数だけ送る。数式は両方に同じテストを置く。

import type { Document, Vec2 } from "./types";

/** 折れ線の分割数の上限(これ以上細かくしても見分けが付かず辺数だけ増える) */
export const MAX_CURVE_SEGMENTS = 200;
/** 既定の許容誤差(紙の長辺=1.0 に対する、曲線と弦の離れ方の上限) */
export const DEFAULT_CURVE_TOL = 0.005;

const EPS = 1e-12;
const TAU = Math.PI * 2;

const sub = (a: Vec2, b: Vec2): Vec2 => [a[0] - b[0], a[1] - b[1]];
const add = (a: Vec2, b: Vec2): Vec2 => [a[0] + b[0], a[1] + b[1]];
const mul = (a: Vec2, k: number): Vec2 => [a[0] * k, a[1] * k];
const len = (a: Vec2): number => Math.hypot(a[0], a[1]);
const cross = (a: Vec2, b: Vec2): number => a[0] * b[1] - a[1] * b[0];
const sq = (a: Vec2): number => a[0] * a[0] + a[1] * a[1];

/** 3点を通る円の中心。3点がほぼ一直線ならnull */
export function circumcenter(a: Vec2, b: Vec2, c: Vec2): Vec2 | null {
  const d = 2 * (a[0] * (b[1] - c[1]) + b[0] * (c[1] - a[1]) + c[0] * (a[1] - b[1]));
  if (Math.abs(d) < EPS) return null;
  const [qa, qb, qc] = [sq(a), sq(b), sq(c)];
  return [
    (qa * (b[1] - c[1]) + qb * (c[1] - a[1]) + qc * (a[1] - b[1])) / d,
    (qa * (c[0] - b[0]) + qb * (a[0] - c[0]) + qc * (b[0] - a[0])) / d,
  ];
}

/** 円弧を折れ線にするときの分割数(1区間の弦の膨らみが tol 以下になる最小の数) */
export function arcSegmentCount(radius: number, sweep: number, tol: number): number {
  const t = Math.max(tol, 1e-9);
  if (radius <= t) return 1;
  const step = 2 * Math.acos(Math.min(1, Math.max(-1, 1 - t / radius)));
  if (step <= EPS) return MAX_CURVE_SEGMENTS;
  return Math.min(MAX_CURVE_SEGMENTS, Math.max(1, Math.ceil(Math.abs(sweep) / step)));
}

const mod = (a: number, m: number): number => ((a % m) + m) % m;

/**
 * 始点・通過点・終点で決まる円弧の折れ線(端点を含む)。
 * segmentsを指定するとその数で等分し、なければ tol から自動で決める。
 * 3点が一直線・退化していればただの線分を返す。
 */
export function arcPolyline(
  p0: Vec2,
  through: Vec2,
  p1: Vec2,
  tol: number,
  segments?: number,
): Vec2[] {
  const c = circumcenter(p0, through, p1);
  if (!c) return [p0, p1];
  const r = len(sub(p0, c));
  if (r < EPS) return [p0, p1];
  const ang = (p: Vec2) => Math.atan2(p[1] - c[1], p[0] - c[0]);
  const a0 = ang(p0);
  const d1 = mod(ang(p1) - a0, TAU);
  // 通過点が始点から見て終点より手前にあるなら反時計回り、そうでなければ時計回り
  const sweep = mod(ang(through) - a0, TAU) < d1 ? d1 : d1 - TAU;
  const n =
    segments === undefined
      ? arcSegmentCount(r, sweep, tol)
      : Math.min(MAX_CURVE_SEGMENTS, Math.max(1, Math.round(segments)));
  const pts: Vec2[] = [];
  for (let i = 0; i <= n; i++) {
    const t = a0 + (sweep * i) / n;
    pts.push([c[0] + r * Math.cos(t), c[1] + r * Math.sin(t)]);
  }
  // 端点は丸め誤差を入れず指定値そのものにする(既存頂点へ確実に吸着させる)
  pts[0] = p0;
  pts[n] = p1;
  return pts;
}

/** 3次ベジェを折れ線にするときの分割数(誤差は max|B''|/(8n²) 以下) */
export function cubicSegmentCount(
  p0: Vec2,
  c1: Vec2,
  c2: Vec2,
  p1: Vec2,
  tol: number,
): number {
  const t = Math.max(tol, 1e-9);
  // 制御点が始点と終点を結ぶ直線に乗っていれば曲線もその直線上にある
  const chord = sub(p1, p0);
  if (len(chord) > EPS) {
    const u = mul(chord, 1 / len(chord));
    const off = (p: Vec2) => Math.abs(cross(sub(p, p0), u));
    if (off(c1) <= t && off(c2) <= t) return 1;
  }
  const v1 = add(sub(p0, mul(c1, 2)), c2);
  const v2 = add(sub(c1, mul(c2, 2)), p1);
  const m = 6 * Math.max(len(v1), len(v2));
  if (m <= EPS) return 1;
  return Math.min(MAX_CURVE_SEGMENTS, Math.max(1, Math.ceil(Math.sqrt(m / (8 * t)))));
}

/** 3次ベジェの折れ線(端点を含む)。S字も引けるので自由度が高い */
export function cubicPolyline(
  p0: Vec2,
  c1: Vec2,
  c2: Vec2,
  p1: Vec2,
  tol: number,
  segments?: number,
): Vec2[] {
  const n =
    segments === undefined
      ? cubicSegmentCount(p0, c1, c2, p1, tol)
      : Math.min(MAX_CURVE_SEGMENTS, Math.max(1, Math.round(segments)));
  const pts: Vec2[] = [];
  for (let i = 0; i <= n; i++) {
    const t = i / n;
    const u = 1 - t;
    pts.push(
      add(
        add(mul(p0, u * u * u), mul(c1, 3 * u * u * t)),
        add(mul(c2, 3 * u * t * t), mul(p1, t * t * t)),
      ),
    );
  }
  pts[0] = p0;
  pts[n] = p1;
  return pts;
}

/** 線分を紙の矩形(0,0)-(w,h)で切り取る(Liang-Barsky法)。掛からなければnull */
function clipToPaper(a: Vec2, d: Vec2, w: number, h: number): [Vec2, Vec2] | null {
  let t0 = -Infinity;
  let t1 = Infinity;
  const limits: [number, number][] = [
    [-d[0], a[0]],
    [d[0], w - a[0]],
    [-d[1], a[1]],
    [d[1], h - a[1]],
  ];
  for (const [p, q] of limits) {
    if (Math.abs(p) < EPS) {
      if (q < 0) return null;
      continue;
    }
    const r = q / p;
    if (p < 0) t0 = Math.max(t0, r);
    else t1 = Math.min(t1, r);
  }
  if (t1 - t0 < EPS) return null;
  return [add(a, mul(d, t0)), add(a, mul(d, t1))];
}

/** 曲線の折り目の両側に入れる「紙が曲がるための線」(ruling)の端点 */
export interface Ruling {
  /** 曲線の上の点(ここから両側へ伸びる) */
  at: Vec2;
  /** へこむ側(曲がりの内側)の端 */
  concave: Vec2;
  /** 膨らむ側(曲がりの外側)の端 */
  convex: Vec2;
}

/**
 * 曲線の折り目の両側に入れる「紙が曲がるための線」。
 *
 * 曲線の折り目は、両側の紙が曲がらないと折れない(平らな板2枚を曲線でつなぐと
 * 角度0以外では紙がちぎれる)。実際の紙では折り目の両側が円錐状に曲がっており、
 * その曲がりを表すのがこの線。折れ線の各内点で、折れ線に直角な向きへ
 * 紙の縁まで伸ばした線を返す。
 */
export function rulingLines(points: Vec2[], paper: Vec2): Ruling[] {
  const out: Ruling[] = [];
  for (let i = 1; i < points.length - 1; i++) {
    const [prev, cur, next] = [points[i - 1], points[i], points[i + 1]];
    const tan = sub(next, prev);
    if (len(tan) < EPS) continue;
    const n = mul([-tan[1], tan[0]], 1 / len(tan));
    const clipped = clipToPaper(cur, n, paper[0], paper[1]);
    if (!clipped) continue;
    const [p, q] = clipped;
    // 左へ曲がる(外積が正)なら、へこむ側は法線の正の向き(=q側)
    const left = cross(sub(cur, prev), sub(next, cur)) > 0;
    const [concave, convex] = left ? [q, p] : [p, q];
    if (len(sub(concave, cur)) < EPS || len(sub(convex, cur)) < EPS) continue;
    out.push({ at: cur, concave, convex });
  }
  return out;
}

/** 2つの線分の交点(交わらなければnull) */
function segIntersection(a: Vec2, b: Vec2, c: Vec2, d: Vec2): Vec2 | null {
  const r = sub(b, a);
  const s = sub(d, c);
  const denom = cross(r, s);
  if (Math.abs(denom) < 1e-15) return null;
  const t = cross(sub(c, a), s) / denom;
  const u = cross(sub(c, a), r) / denom;
  if (t < 0 || t > 1 || u < 0 || u > 1) return null;
  return add(a, mul(r, t));
}

/**
 * fromからtoへ向かって、最初にぶつかる既存の折り目までで切った終点。
 * 曲がるための線が他の折り目を突き抜けて関係のない場所まで伸びるのを防ぐ
 * (実際の紙でも紙が曲がる範囲は隣の折り目までで区切られる)。
 * Rust側の insert_rulings と同じ切り方(crates/ori3-cp/src/curve.rs)。
 */
export function firstCrossing(doc: Document, from: Vec2, to: Vec2): Vec2 {
  const total = len(sub(to, from));
  if (total < EPS) return to;
  const dir = mul(sub(to, from), 1 / total);
  const byId = new Map(doc.cp.vertices.map((v) => [v.id, v.pos]));
  let best = total;
  for (const e of doc.cp.edges) {
    if (e.kind === "Aux") continue;
    const p0 = byId.get(e.v0);
    const p1 = byId.get(e.v1);
    if (!p0 || !p1) continue;
    const q = segIntersection(from, to, p0, p1);
    if (!q) continue;
    const t = sub(q, from)[0] * dir[0] + sub(q, from)[1] * dir[1];
    // 出発点そのもの(曲線の上)での交わりは無視する
    if (t > 1e-6 && t < best) best = t;
  }
  return add(from, mul(dir, best));
}

/** 曲線の描き方(円弧=3点、ベジェ=4点) */
export type CurveShape = "arc" | "bezier";

/** 曲線の折り目の設定(コンテキストパネルで切り替える) */
export interface CurveOptions {
  /** 曲線モードか(オフなら今までどおり2クリックの直線) */
  enabled: boolean;
  shape: CurveShape;
  /** 分割数。null=自動(誤差 DEFAULT_CURVE_TOL 以内になるまで細かくする) */
  segments: number | null;
  /** 紙が曲がるための線も一緒に引くか(これが無いと曲線折りは折れない) */
  rulings: boolean;
}

export const DEFAULT_CURVE: CurveOptions = {
  enabled: false,
  shape: "arc",
  segments: null,
  rulings: true,
};

/** その描き方に必要なクリックの数 */
export const CURVE_STEPS: Record<CurveShape, number> = { arc: 3, bezier: 4 };

export const CURVE_LABEL: Record<CurveShape, string> = {
  arc: "円弧",
  bezier: "ベジェ",
};

/**
 * 集めたクリック(と今のカーソル位置)から曲線の折れ線を作る。
 * 点が足りなければ、足りない分をカーソル位置で補って描いている最中の形を返す。
 * 円弧は[始点, 終点, 通過点]、ベジェは[始点, 終点, 制御点1, 制御点2]の順。
 */
export function curvePolyline(
  shape: CurveShape,
  points: Vec2[],
  opts: { segments: number | null; tol?: number },
): Vec2[] | null {
  const tol = opts.tol ?? DEFAULT_CURVE_TOL;
  const n = opts.segments ?? undefined;
  if (points.length < 2) return null;
  if (shape === "arc") {
    const [p0, p1, through] = points;
    if (through === undefined) return [p0, p1];
    return arcPolyline(p0, through, p1, tol, n);
  }
  const [p0, p1, c1, c2] = points;
  if (c1 === undefined) return [p0, p1];
  return cubicPolyline(p0, c1, c2 ?? c1, p1, tol, n);
}

/** 次に何をすればよいかの案内(1行)。doneは済んだクリックの数 */
export function curveHint(shape: CurveShape, done: number, rulings: boolean): string {
  const steps: Record<CurveShape, string[]> = {
    arc: ["曲線の始点をクリック", "曲線の終点をクリック", "曲線を通したい点をクリック"],
    bezier: [
      "曲線の始点をクリック",
      "曲線の終点をクリック",
      "始点側の引っぱり先をクリック",
      "終点側の引っぱり先をクリック",
    ],
  };
  const list = steps[shape];
  const text = list[Math.min(done, list.length - 1)];
  const extra = rulings ? "" : "(曲がるための線なし)";
  return `${CURVE_LABEL[shape]}${extra}: ${text}(Escで中止)`;
}

File: crates/ori3-cp/src/curve.rs
//! 曲線の折り目(CPE-011): 円弧・3次ベジェを、指定した誤差以内の折れ線にする。
//!
//! 展開図のデータ構造は直線の辺だけなので、曲線は「十分細かい折れ線」として
//! 入れる。実際の紙でも曲線折りは連続した曲がりなので、分割を細かくすれば
//! 見た目・折れ方ともに近づく(残る差はたわみ計算 `ori3-soft` が埋める)。
//!
//! 同じ計算は画面側(apps/desktop/src/lib/curve.ts)にもある。Tauriコマンドは
//! 増やさない約束なので、描いている最中の形は画面側で作り、確定したときに
//! 既存の `AddSegment` を折れ線の本数だけ送る。数式は両方に同じテストを置く。

use std::f64::consts::TAU;

use glam::DVec2;
use ori3_model::{CreasePattern, EdgeId, EdgeKind};

use crate::graph::insert_segment;

/// 折れ線の分割数の上限。これ以上細かくしても画面では見分けが付かず、
/// 辺数だけが増えて操作が重くなる。
pub const MAX_CURVE_SEGMENTS: u32 = 200;
/// 既定の許容誤差(紙の長辺=1.0 に対する、曲線と弦の離れ方の上限)。
pub const DEFAULT_CURVE_TOL: f64 = 0.005;

const EPS: f64 = 1e-12;

/// 3点を通る円の中心。3点がほぼ一直線ならNone。
fn circumcenter(a: DVec2, b: DVec2, c: DVec2) -> Option<DVec2> {
    let d = 2.0 * (a.x * (b.y - c.y) + b.x * (c.y - a.y) + c.x * (a.y - b.y));
    if d.abs() < EPS {
        return None;
    }
    let (qa, qb, qc) = (a.length_squared(), b.length_squared(), c.length_squared());
    Some(DVec2::new(
        (qa * (b.y - c.y) + qb * (c.y - a.y) + qc * (a.y - b.y)) / d,
        (qa * (c.x - b.x) + qb * (a.x - c.x) + qc * (b.x - a.x)) / d,
    ))
}

/// 円弧を折れ線にするときの分割数。1区間の弦の膨らみ r(1−cos(Δ/2)) が
/// tol 以下になる最小の数を返す。
#[must_use]
pub fn arc_segment_count(radius: f64, sweep: f64, tol: f64) -> u32 {
    let tol = tol.max(1e-9);
    if radius <= tol {
        return 1;
    }
    let step = 2.0 * (1.0 - tol / radius).clamp(-1.0, 1.0).acos();
    if step <= EPS {
        return MAX_CURVE_SEGMENTS;
    }
    ((sweep.abs() / step).ceil() as u32).clamp(1, MAX_CURVE_SEGMENTS)
}

/// 始点・通過点・終点で決まる円弧の折れ線(端点を含む)。
/// `segments` を指定するとその数で等分し、Noneなら `tol` から自動で決める。
/// 3点が一直線・退化していればただの線分を返す。
#[must_use]
pub fn arc_polyline(
    p0: [f64; 2],
    through: [f64; 2],
    p1: [f64; 2],
    tol: f64,
    segments: Option<u32>,
) -> Vec<[f64; 2]> {
    let (a, m, b) = (DVec2::from(p0), DVec2::from(through), DVec2::from(p1));
    let Some(c) = circumcenter(a, m, b) else {
        return vec![p0, p1];
    };
    let r = (a - c).length();
    if r < EPS {
        return vec![p0, p1];
    }
    let ang = |p: DVec2| (p - c).y.atan2((p - c).x);
    let (a0, am, a1) = (ang(a), ang(m), ang(b));
    // 通過点が始点から見て終点より手前にあるなら反時計回り、そうでなければ時計回り
    let d1 = (a1 - a0).rem_euclid(TAU);
    let sweep = if (am - a0).rem_euclid(TAU) < d1 {
        d1
    } else {
        d1 - TAU
    };
    let n = segments.map_or_else(
        || arc_segment_count(r, sweep, tol),
        |n| n.clamp(1, MAX_CURVE_SEGMENTS),
    );
    let mut pts: Vec<[f64; 2]> = (0..=n)
        .map(|i| {
            let t = a0 + sweep * f64::from(i) / f64::from(n);
            [c.x + r * t.cos(), c.y + r * t.sin()]
        })
        .collect();
    // 端点は丸め誤差を入れず指定値そのものにする(既存頂点へ確実に吸着させる)
    pts[0] = p0;
    pts[n as usize] = p1;
    pts
}

/// 3次ベジェを折れ線にするときの分割数。
/// 折れ線の誤差は max|B''| / (8n²) 以下なので、それが tol 以下になる数を返す。
#[must_use]
pub fn cubic_segment_count(
    p0: [f64; 2],
    c1: [f64; 2],
    c2: [f64; 2],
    p1: [f64; 2],
    tol: f64,
) -> u32 {
    let tol = tol.max(1e-9);
    let (p0, c1, c2, p1) = (
        DVec2::from(p0),
        DVec2::from(c1),
        DVec2::from(c2),
        DVec2::from(p1),
    );
    // 制御点が始点と終点を結ぶ直線に乗っていれば曲線もその直線上にある
    let chord = p1 - p0;
    if chord.length() > EPS {
        let u = chord.normalize();
        let off = |p: DVec2| (p - p0).perp_dot(u).abs();
        if off(c1) <= tol && off(c2) <= tol {
            return 1;
        }
    }
    let m = 6.0
        * (p0 - 2.0 * c1 + c2)
            .length()
            .max((c1 - 2.0 * c2 + p1).length());
    if m <= EPS {
        return 1;
    }
    ((m / (8.0 * tol)).sqrt().ceil() as u32).clamp(1, MAX_CURVE_SEGMENTS)
}

/// 3次ベジェの折れ線(端点を含む)。S字も引けるので自由度が高い。
#[must_use]
pub fn cubic_polyline(
    p0: [f64; 2],
    c1: [f64; 2],
    c2: [f64; 2],
    p1: [f64; 2],
    tol: f64,
    segments: Option<u32>,
) -> Vec<[f64; 2]> {
    let n = segments.map_or_else(
        || cubic_segment_count(p0, c1, c2, p1, tol),
        |n| n.clamp(1, MAX_CURVE_SEGMENTS),
    );
    let (a, b, c, d) = (
        DVec2::from(p0),
        DVec2::from(c1),
        DVec2::from(c2),
        DVec2::from(p1),
    );
    let mut pts: Vec<[f64; 2]> = (0..=n)
        .map(|i| {
            let t = f64::from(i) / f64::from(n);
            let u = 1.0 - t;
            let p =
                a * (u * u * u) + b * (3.0 * u * u * t) + c * (3.0 * u * t * t) + d * (t * t * t);
            [p.x, p.y]
        })
        .collect();
    pts[0] = p0;
    pts[n as usize] = p1;
    pts
}

/// 線分を紙の矩形(0,0)-(w,h)で切り取る(Liang-Barsky法)。掛からなければNone。
fn clip_to_paper(a: DVec2, d: DVec2, w: f64, h: f64) -> Option<(DVec2, DVec2)> {
    let (mut t0, mut t1) = (f64::NEG_INFINITY, f64::INFINITY);
    for (p, q) in [(-d.x, a.x), (d.x, w - a.x), (-d.y, a.y), (d.y, h - a.y)] {
        if p.abs() < EPS {
            if q < 0.0 {
                return None;
            }
            continue;
        }
        let r = q / p;
        if p < 0.0 {
            t0 = t0.max(r);
        } else {
            t1 = t1.min(r);
        }
    }
    if t1 - t0 < EPS {
        return None;
    }
    Some((a + d * t0, a + d * t1))
}

/// 曲線の折り目の両側に入れる「紙が曲がるための線」(ruling)。
///
/// 曲線の折り目は、両側の紙が曲がらないと折れない(平らな板2枚を曲線で
/// つなぐと、角度0以外では紙がちぎれる)。実際の紙では折り目の両側が
/// 円錐状に曲がっており、その曲がりを表すのがこの線。
///
/// 折れ線の各内点で、折れ線に直角な向きへ紙の縁まで伸ばした線を返す。
/// 各要素は `[へこむ側の端, 折れ線上の点, 膨らむ側の端]`(向きは曲がる向きで決まる)。
#[must_use]
pub fn ruling_lines(points: &[[f64; 2]], paper: [f64; 2]) -> Vec<[[f64; 2]; 3]> {
    let mut out = Vec::new();
    for i in 1..points.len().saturating_sub(1) {
        let (prev, cur, next) = (
            DVec2::from(points[i - 1]),
            DVec2::from(points[i]),
            DVec2::from(points[i + 1]),
        );
        let tan = next - prev;
        if tan.length() < EPS {
            continue;
        }
        let n = DVec2::new(-tan.y, tan.x).normalize();
        let Some((p, q)) = clip_to_paper(cur, n, paper[0], paper[1]) else {
            continue;
        };
        // 左へ曲がる(外積が正)なら、へこむ側は法線の正の向き(=q側)
        let left = (cur - prev).perp_dot(next - cur) > 0.0;
        let (concave, convex) = if left { (q, p) } else { (p, q) };
        if (concave - cur).length() < EPS || (convex - cur).length() < EPS {
            continue;
        }
        out.push([concave.into(), cur.into(), convex.into()]);
    }
    out
}

/// 折れ線を展開図へ入れる(区間ごとに `insert_segment`)。
/// 戻り値は新しくできた辺のID(後の区間の挿入で再分割され得る点は同じ)。
pub fn insert_polyline(cp: &mut CreasePattern, points: &[[f64; 2]], kind: EdgeKind) -> Vec<EdgeId> {
    points
        .windows(2)
        .flat_map(|w| insert_segment(cp, w[0], w[1], kind))
        .collect()
}

/// 曲がるための線を展開図へ入れる(先に `insert_polyline` で曲線を入れておく)。
/// 折り目の両側で曲がる向きが逆になるので、へこむ側は曲線と反対の線種、
/// 膨らむ側は同じ線種にする(画面の山谷の色を実際の曲がりに合わせる)。
pub fn insert_rulings(
    cp: &mut CreasePattern,
    points: &[[f64; 2]],
    paper: [f64; 2],
    curve_kind: EdgeKind,
) -> Vec<EdgeId> {
    let opposite = if curve_kind == EdgeKind::Mountain {
        EdgeKind::Valley
    } else {
        EdgeKind::Mountain
    };
    let mut ids = Vec::new();
    for [concave, cur, convex] in ruling_lines(points, paper) {
        ids.extend(insert_segment(
            cp,
            cur,
            first_crossing(cp, cur, concave),
            opposite,
        ));
        ids.extend(insert_segment(
            cp,
            cur,
            first_crossing(cp, cur, convex),
            curve_kind,
        ));
    }
    ids
}

/// `from`から`to`へ向かって、最初にぶつかる既存の折り目までで切った終点。
/// ぶつからなければ`to`のまま。曲がるための線が他の折り目を突き抜けて
/// 関係のない場所まで伸びるのを防ぐ(実際の紙でも紙が曲がる範囲は
/// 隣の折り目までで区切られる)。
fn first_crossing(cp: &CreasePattern, from: [f64; 2], to: [f64; 2]) -> [f64; 2] {
    let (a, b) = (DVec2::from(from), DVec2::from(to));
    let len = (b - a).length();
    if len < EPS {
        return to;
    }
    let dir = (b - a) / len;
    let vpos: std::collections::BTreeMap<_, _> = cp
        .vertices
        .iter()
        .map(|v| (v.id, DVec2::from(v.pos)))
        .collect();
    let mut best = len;
    for e in &cp.edges {
        if e.kind == EdgeKind::Aux {
            continue;
        }
        let (Some(&p0), Some(&p1)) = (vpos.get(&e.v0), vpos.get(&e.v1)) else {
            continue;
        };
        if let Some(q) = ori3_geometry::seg_intersection(a, b, p0, p1) {
            let t = (q - a).dot(dir);
            // 出発点そのもの(曲線の上)での交わりは無視する
            if t > 1e-6 && t < best {
                best = t;
            }
        }
    }
    (a + dir * best).into()
}
